### ISSUE TITLE
Region add/remove issues

### ISSUE THREAD

**Issue by eamodio** [naver/billboard.js#3037]
If I set the `chart.regions` directly things seem to work fine, but I'm trying to dynamically add/remove "classes" of regions.

I am using regions as both "markers" and shaded regions and I have multiple different types of regions each with their own "class".

When I try to update a set of regions for a single "class" using `chart.regions.remove` and then `chart.regions.add` things don't seem to work properly. Sometimes it seems like the `remove` call removes markers that don't have that class, but then other times seems to fail to remove even markers of that class as I see them keep overlaying on top of one another.

Here is an example of my code:
```ts
chart.regions.remove({ class: 'visible-area' });
if (this.visibleDays == null) return;

chart.regions.add({
	axis: 'x',
	start: this.visibleDays.bottom,
	end: this.visibleDays.top,
	class: 'visible-area',
} satisfies RegionOptions);
```

Should this work? Or am I not understanding how the remove/add calls are supposed to work?

**Comment by netil** [comment]
@eamodio, note that `.regions.remove()` argument type isn't correct in your example.

The key name should be `classes`, not `class` and the key value also should be an array.
```js
chart.regions.remove({
    classes: ['visible-area']
});
```

So, from the `chart.regions.remove({ class: 'visible-area' });` call, is considered no valid param is given. 
Hence all regions will be removed.

checkout the example:
- https://jsfiddle.net/netil/tm6wea8h/

**Comment by eamodio** [comment]
@netil Thanks -- that seems to fix a lot of the issues I was seeing. 

It looks like the typings are wrong for the `regions.remove`:
https://github.com/naver/billboard.js/blob/master/types/chart.d.ts#L100-L119

I think the issue I am still seeing is related to the animation when a region comes into view. Is there a way to turn that off?

I am trying to update a region while scrolling another list, and it seems to kind of hide and show the region almost sporadically

![regions](https://user-images.githubusercontent.com/641685/212629820-137a5f11-083b-48c8-a41e-18800b614b1a.gif)

Here I am scrolling down a line at a time, and it should be updating the region on each one, but something odd is happening.

**Comment by netil** [comment]
@eamodio, thanks for inform the incorrect type definition.

> It looks like the typings are wrong for the regions.remove:
> https://github.com/naver/billboard.js/blob/master/types/chart.d.ts#L100-L119

Will be fixed ASAP.

> I think the issue I am still seeing is related to the animation when a region comes into view. Is there a way to turn that off?

The easiest way is turning off the transition by specifying `transition.duration=0`, but it will affect entire chart updates.
Checkout the below example.
- https://jsfiddle.net/netil/tm6wea8h/2/

Just turning off on `grid`, it should probably add some new option to grid API to enable/disable the transition.

**Comment by eamodio** [comment]
@netil Thanks! Turning the transition off did the trick. Yeah, it would be nice for finer-grained control on that if possible (e.g. just for the regions).

I also don't know if it would help optimize layout/repaint, but it could be nice to have a `regions.update()` method, that combines both add and remove

**Comment by eamodio** [comment]
@netil Thanks for the typing updates. Though I noticed an issue, with the new changes:

The `RegionOptions` in chart.d.ts doesn't match the new options.

```ts
export interface RegionOptions {
	axis?: string;
	start?: string | number | Date;
	end?: string | number | Date;
	class?: string;
}
```

```ts
	regions: {
		/**
		 * Update regions.
		 * @param regions Regions will be replaced with this argument. The format of this argument is the same as regions.
		 */
		(regions: Array<{
			axis: "x" | "y" | "y2",
			start?: number | Date,
			end?: number | Date,
			class?: string
		}>): void;

		/**
		 * Add new region. This API adds new region instead of replacing like regions.
		 * @param grids New region will be added. The format of this argument is the same as regions and it's possible to give an Object if only one region will be added.
		 */
		add<T = {
			axis: "x" | "y" | "y2",
			start?: number | Date,
			end?: number | Date,
			class?: string
		}>(regions: T | T[]): void;

		/**
		 * Remove regions. This API removes regions.
		 * @param args This argument should include classes. If classes is given, the regions that have one of the specified classes will be removed. If args is not given, all of regions will be
		 * removed.
		 */
		remove(args?: { classes: string[] }): void;
	};
```

**Comment by netil** [comment]
@eamodio thanks for reporting.

**Comment by eamodio** [comment]
Related to my comment here: https://github.com/naver/billboard.js/issues/3037#issuecomment-1384821188

Thoughts on a `regions.update()` method? I am seeing some performance lag/issues since I am trying to update the regions very frequently.

My use-case is here: https://github.com/gitkraken/vscode-gitlens/discussions/2477

**Comment by netil** [comment]
@eamodio 
> Thoughts on a regions.update() method? I am seeing some performance lag/issues since I am trying to update the regions very frequently.

Currently, when call grid APIs it goes redrawing the chart. I'll be refactoring only to perform on grid related elements.

**Comment by eamodio** [comment]
Sweet - looking forward to checking it out!

**Comment by netil** [comment]
@eamodio you can use `nightly`, which is the latest build from the master branch and the purpose is to give chance to test before the official release.
- https://github.com/naver/billboard.js#nightly-version

**Comment by eamodio** [comment]
@netil I've tried out the latest version (`3.8.0-next.1`) and things look pretty good, but I am seeing an issue where when I update a particular region, all the other regions get an extra `<rect>` element added to them

![image](https://user-images.githubusercontent.com/641685/229992434-eabb97b6-fdbe-4508-b9cc-4144d2b63ad5.png)

In the above case, I was updating the `visible-area` region, by calling `regions.remove({ classes: ['visible-area'] })` then `regions.add()` with the new `visible-area` regions -- as can be seen [here](https://github.com/gitkraken/vscode-gitlens/blob/61dca228e66b52068f981f7c3ac67e4168376e04/src/webviews/apps/plus/graph/minimap/minimap.ts#L449-L452)

And because I'm updating those visible regions on scroll, it generates LOTS of those extra `<rect>` elements which slows things down pretty quickly.

**Comment by netil** [comment]
@eamodio confirmed the issue. Seems having some side-effect from the #3147 issue.

**Comment by netil** [comment]
released [`3.8.0-next.2`](https://github.com/naver/billboard.js/releases/tag/3.8.0-next.2).

**Comment by eamodio** [comment]
@netil looks great! TY
